### PR TITLE
Macsec High level design

### PR DESCRIPTION
**Pull request by Pterosaur**
| PR title                                                                                             | state                                                                                                               | context                                                                                                                  |
| ---------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------ |
| [Add MACsec schema](https://github.com/Azure/sonic-swss-common/pull/403)                             | GitHub issue/pull request detail   | GitHub pull request check contexts   |
| [Add wpa supplicant build tool](https://github.com/Azure/sonic-build-tools/pull/193)                             | GitHub issue/pull request detail   | GitHub pull request check contexts   |
| [MACsec container and wpa_supplicant component](https://github.com/Azure/sonic-buildimage/pull/5700) | GitHub issue/pull request detail   | GitHub pull request check contexts   |
| [Add debian rule for wpa_supplicant](https://github.com/Azure/sonic-wpa-supplicant/pull/18)             | GitHub issue/pull request detail | GitHub pull request check contexts |
| [[SONiC plugin]: Add SONiC WPA_Supplicant Plugin ](https://github.com/Azure/sonic-wpa-supplicant/pull/22)             | GitHub issue/pull request detail | GitHub pull request check contexts |
| [Move SSCI from SC to SA and change packet number field name to adapt sai 1.7.1 ](https://github.com/Azure/sonic-wpa-supplicant/pull/26)             | GitHub issue/pull request detail | GitHub pull request check contexts |
| [Add utility for string and redis](https://github.com/Azure/sonic-swss-common/pull/434)                     | GitHub issue/pull request detail         | GitHub pull request check contexts         |
| [[MACsecMgr]: Add MACsec Manager](https://github.com/Azure/sonic-swss/pull/1475)                     | GitHub issue/pull request detail         | GitHub pull request check contexts         |
| [[orchagent]: Add MACsec Orchagent](https://github.com/Azure/sonic-swss/pull/1474)                   | GitHub issue/pull request detail         | GitHub pull request check contexts         |
| [[vslib]Add MACsec Filters](https://github.com/Azure/sonic-sairedis/pull/713)                               | GitHub issue/pull request detail      | GitHub pull request check contexts      |
| [[vslib]Add MACsec Forwarder](https://github.com/Azure/sonic-sairedis/pull/714)                               | GitHub issue/pull request detail      | GitHub pull request check contexts      |
| [[vslib]Add MACsec Manager](https://github.com/Azure/sonic-sairedis/pull/715)                               | GitHub issue/pull request detail      | GitHub pull request check contexts      |
| [[vslib]Add helper functions, findObjects and dumpObject](https://github.com/Azure/sonic-sairedis/pull/716)                               | GitHub issue/pull request detail      | GitHub pull request check contexts      |
| [[vslib] Add StateBase function for MACsec](https://github.com/Azure/sonic-sairedis/pull/717)                               | GitHub issue/pull request detail      | GitHub pull request check contexts      |
| [[vslib]Add MACsec forward and filters to HostInterfaceInfo](https://github.com/Azure/sonic-sairedis/pull/719)                               | GitHub issue/pull request detail      | GitHub pull request check contexts      |
| [Add MACsec meta methods](https://github.com/Azure/sonic-sairedis/pull/704)                          | GitHub issue/pull request detail      | GitHub pull request check contexts      |
| [Add FlexCounter for MACsec SA](https://github.com/Azure/sonic-sairedis/pull/684)                    | GitHub issue/pull request detail      | GitHub pull request check contexts      |
| [[vslib]: Add MACsec state to state base](https://github.com/Azure/sonic-sairedis/pull/722)                               | GitHub issue/pull request detail      | GitHub pull request check contexts      |
| [[vslib]: adapt macsec sai 1.7.1](https://github.com/Azure/sonic-sairedis/pull/755)                               | GitHub issue/pull request detail      | GitHub pull request check contexts      |

### CODE REVIEW

File: doc/macsec/MACsec_hld.md [hunk 1]
```diff
@@ -0,0 +1,613 @@
...
+#### 3.3.2 MACsec Egress SC Table
+
+``` rfc5234
+"MACSEC_EGRESS_SC"|{{port_name}}|{{SCI}}
+    "state":{{OK}}
+
+; Defines schema for MACsec Egress SC table attributes
+key          = MACSEC_EGRESS_SC:port_name:sci     ; MACsec SC identifier
+; field      = value
+state        = "" / "ok"                          ; The MACsec SC has been created
```
Comment (prsunny) [inline doc/macsec/MACsec_hld.md:321]: I don't think "" is a value for state. Either it is present with "ok" or the table itself is not present. Can you check and correct here and the below ones?

File: doc/macsec/MACsec_hld.md [hunk 2]
```diff
@@ -0,0 +1,613 @@
...
+|         PORT          |     N      | The specified port name.                                                                                                                                                            |
+|       CONFNAME        |     N      | Config file path. When using MACsec, the following parameters should be set <br> eapol_version=3 <br> ap_scan=0                                                                     |
+|      NETWORK_ID       |     N      | The Identifier of network at this port                                                                                                                                              |
+|          CAK          |     N      | The Secure Connectivity Association Key. <br> This value is the value PRIMARY_CAK in MACsec profile table of CONFIG DB.                                                             |
+|          CAN          |     N      | The CAK Name. <br> This value is the value PRIMARY_CAN in MACsec profile table of CONFIG DB.                                                                                        |
```
Comment (prsunny) [inline doc/macsec/MACsec_hld.md:384]: CKN?

File: doc/macsec/MACsec_hld.md [hunk 3]
```diff
@@ -0,0 +1,613 @@
...
+# Disable MACsec driver at the port
+wpa_cli -g{{DOMAIN_SOCK}} interface_remove {{PORT}}
+```
+
+#### 3.4.2 WPA Supplicant
+
+Wpa_supplicant(<https://salsa.debian.org/debian/wpa>) is Wi-Fi Protected Access client and IEEE 802.1X(MACsec) supplicant. A single instance of wpa_supplicant is able to manage multiple MACsec channels. Wpa_supplicant will be managed by supervisor in the macsec container as a daemon. Start command is `wpa_supplicant -Dmacsec_sonic -g{{DOMAIN_SOCK}}`.
+The first parameter, macsec_sonic, indicates that the SONiC MACsec plugin is used.This parameter should be declared at SONiC MACsec Plugin.
+The second parameter, DOMAIN_SOCK, indicates the position of domain sock, that will be used to interact with wpa_cli.
+The details of SONiC MACsec plugin will be mentioned at next chapter.
```
Comment (prsunny) [inline doc/macsec/MACsec_hld.md:446]: what is referring to next chapter?, do you mean, next phase?

File: doc/macsec/MACsec_hld.md [hunk 4]
```diff
@@ -0,0 +1,613 @@
...
+
+##### 3.4.2.1 Extension Parameters
+
+The following parameters should be extended in WPA_SUPPLICANT for support SONiC MACsec requirements.
```
Comment (prsunny) [inline doc/macsec/MACsec_hld.md:450]: to support

File: doc/macsec/MACsec_hld.md [hunk 5]
```diff
@@ -0,0 +1,613 @@
...
+
+- Delete SC
+  1. Monitor the DEL message from the MACsec Ingress/Egress SC Table in APP DB
+  2. Collect MACsec SC and flow stats
+  3. Delete Flex counter of MACsec SC
+  4. Delete the ingress/egress MACsec SC
+  5. Delete the ingress/egress MACsec SA
+
+- Create Ingress SA
+  1. Monitor the DEL message from the MACsec  Ingress SA Table
```
Comment (prsunny) [inline doc/macsec/MACsec_hld.md:562]: SET message instead of DEL

File: doc/macsec/MACsec_hld.md [hunk 6]
```diff
@@ -0,0 +1,613 @@
...
+
+- Create Egress SA
+  1. Monitor the SET message from the MACsec Egress SC Table, only if ENCODING_AN change
+  2. The ACTIVE filed should be true, otherwise exit this process
+  3. Create egress MACsec egress SA
+  4. Set Flex Counter of MACsec SA stats
+
+- Delete SA
+  1. Monitor the DEL message from the MACsec SA Table
+  2. Collect SA Stats
```
Comment (prsunny) [inline doc/macsec/MACsec_hld.md:575]: Is this intended in Delete SA?

File: doc/macsec/MACsec_hld.md [hunk 7]
```diff
@@ -0,0 +1,613 @@
...
+  3. Delete Flex counter of MACsec SA stats
+  4. Delete MACsec SA
+
+##### 3.4.4.2 Flex Counter Interval
+
+Wpa_supplicant need to monitor the packet number for SAK refreshing. But if a copy of packet number delayed more than the preparation time of SAK, the requirement of SAK refreshing may not be realized by wpa_supplicant, which will cause the packet number to be exhausted.
+
+- MPN=maximal packet number, which indicates the maximal packet number, it should be 4,294,967,295 if packet number is 32bit
```
Comment (prsunny) [inline doc/macsec/MACsec_hld.md:583]: maximum?

File: doc/macsec/MACsec_hld.md [hunk 8]
```diff
@@ -0,0 +1,613 @@
...
+- RT=refresh threshold, which indicates that the SAK should be refreshed if the packet number increases to a threshold. This number is about 75% of MPN.
+- MPB=maximal port bandwidth, which indicates the maximal bandwidth at the port
+- MMPS=minimal MACsec packet size, which indicates the minimal packet size of MACsec, it should be 44 bytes
```
Comment (prsunny) [inline doc/macsec/MACsec_hld.md:586]: minimum?

File: doc/macsec/MACsec_hld.md [hunk 9]
```diff
@@ -0,0 +1,613 @@
...
+
+### 1.1 Functional requirements
+
+This section describes the SONiC requirements for MACsec primary feature.
+
+At a high level the following should be supported:
+
+#### Phase I
+
+- MACsec should be supported on physical port
```
Comment (prsunny) [inline doc/macsec/MACsec_hld.md:83]: May be we should also provide a reference to PortChannels

File: doc/macsec/MACsec_hld.md [hunk 10]
```diff
@@ -0,0 +1,613 @@
...
+- Choose the cipher suite by the wpa_cli
+- A part of parameters of wpa_supplicant can be updated on the fly
+
+#### Phase III
+
+- CLI commands to configure MACsec
+
+## 2 Architecture Design
+
+This chapter shows the MACsec interface stack of virtual switch and real switch.
```
Comment (prsunny) [inline doc/macsec/MACsec_hld.md:101]: Some of the container parts are missing in design. We should provide some details on the following:
1. When will the container be running
2. Is the FEATURE disabled/enabled by default

File: doc/macsec/MACsec_hld.md [hunk 11]
```diff
@@ -0,0 +1,613 @@
...
+- Support Cipher: GCM-AES-128 and GCM-AES-256
+- Secure Association Key(SAK) can be replaced without service outage
+
+#### Phase II
+
+- MACsec can support  Extension packet number(XPN), which means to support Cipher Suites: GCM-AES-XPN-128 and GCM-AES-XPN-256
+- SAK can be refreshed proactively.
+- Primary and Fallback secure Connectivity Association Key can be supported simultaneously.
+- Choose the cipher suite by the wpa_cli
+- A part of parameters of wpa_supplicant can be updated on the fly
```
Comment (lguohan) [inline doc/macsec/MACsec_hld.md:93]: can you specify the parameter list?

File: doc/macsec/MACsec_hld.md [hunk 12]
```diff
@@ -0,0 +1,613 @@
...
+
+#### Phase III
+
+- CLI commands to configure MACsec
```
[the remark this answers is not in the document]
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md:104]: Done, please review it.

File: doc/macsec/MACsec_hld.md [hunk 13]
```diff
@@ -0,0 +1,613 @@
...
+
+This document provides general information about the MACsec feature implementation in SONiC.
+
+## Abbreviation
+
+| Abbreviation | Description                              |
+| ------------ | ---------------------------------------- |
+| CA           | Secure Connectivity Association          |
+| CAK          | Secure Connectivity Association Key      |
+| CAN          | Secure Connectivity Association Key Name |
```
[the remark this answers is not in the document]
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md]: Done, please review it.

File: doc/macsec/MACsec_hld.md (hunk 10 above)
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md:108]: When will the container be running
Add the information at : https://github.com/Azure/SONiC/pull/652/files#diff-600ad1f148d06a6c05f027cd045d6c4fR122
Is the FEATURE disabled/enabled by default
Add the information at : https://github.com/Azure/SONiC/pull/652/files#diff-600ad1f148d06a6c05f027cd045d6c4fR198-R199

File: doc/macsec/MACsec_hld.md [hunk 14]
```diff
@@ -0,0 +1,613 @@
...
+                                                       ; The cipher suite for MACsec.
+ENABLE_PROTECT              = "true" / "false"         ; Whether protect the traffic
+ENABLE_ENCRYPT              = "true" / "false"         ; Whether encrypt the traffic
+ENABLE_REPLAY_PROTECT       = "true" / "false"         ; Whether enable replay protect.
+REPLAY_WINDOW               = DIGITS                   ; Replay window size that is the number of
+                                                       ; packets that could be out of order.
+SEND_SCI                    = "true" / "false"         ; Whether send SCI.
+```
+
+#### 3.2.2 MACsec Egress SC Table
```
[the remark this answers is not in the document]
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md:237]: As we discussion, the naming convention, we will follow the MACsec SAI's.

File: doc/macsec/MACsec_hld.md [hunk 15]
```diff
@@ -0,0 +1,613 @@
...
+
+- CLI commands to configure MACsec
+
+## 2 Architecture Design
+
+This chapter shows the MACsec interface stack of virtual switch and real switch.
+
+Virtual switch use the Linux MACsec driver to support the functionality of MACsec and the MACsec interface is imposed on Ethernet interface.
+
+Real switch use the ASIC chip as the MACsec Security Entity(SecY) which will be imposed on physic interface. And the ethernet port will be above the SecY.
```
[the remark this answers is not in the document]
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md]: Done, please review it.

File: doc/macsec/MACsec_hld.md [hunk 16]
```diff
@@ -0,0 +1,613 @@
...
+
+In all scenarios, both virtual and real switch, the IP address will be assigned to the Port. All traffic, except EAPOL packets,  on the Port will be encrypted and then these traffic will be sent by Ethernet or Physical Interface. Meanwhile, all traffic, except EAPOL packets, on the physical port will be validated and decrypted and then these traffic will be delivered to Port or discarded if the validation fails.
```
[the remark this answers is not in the document]
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md]: Done, please review it.

File: doc/macsec/MACsec_hld.md [hunk 17]
```diff
@@ -0,0 +1,613 @@
...
+
+![interface stack](images/interface_stack.png)  
+
+## 3 Modules Design
+
+The following figure depicts the data flow and related components of MACsec. Different colored boxes indicate the components that belong to different containers or spaces. The following paragraph only lists the necessary components that are needed by the functionality of MACsec.
+
+- The orange means these components belong to MACsec container. This container include the control plane of MACsec that include MACsec enable/disable, MACsec session negotiation and key exchange. This container consists of daemons, MACsecMgr and wpa_supplicant, and a tool, wpa_cli.
+  - **MACsecMgr** needs to control the availability of MACsec on specified interfaces by using wpa_cli commands to communicate with wpa_supplicant daemon.
```
[the remark this answers is not in the document]
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md]: Done, please review it.

File: doc/macsec/MACsec_hld.md [hunk 18]
```diff
@@ -0,0 +1,613 @@
...
+  - **wpa_supplicant** uses the MKA protocol to manage the MACsec session and key exchange in MACsec control plane. It calls the predefined MACsec plugin APIs to communicate with SONiC MACsec plugin.
+  - **SONiC** MACsec Plugin is a plugin of wpa_supplicant that follows the predefined APIs of wpa_supplicant. It is responsible for bi-directional conversion of the MACsec parameters and SONiC database entries.
+
+- The green means these components are in SWSS container. This container uses the SAI APIs to control the MACsec security entities(SecY) according to databases entries and to synchronize the statistics from SecY to COUNTER_DB.
+  - **MACsecOrch** is a module of orchagent, that uses SAI APIs to manage the SecY according to messages from databases and synchronized the statistics of SecY to COUNTER_DB.
+
+- The blue one is (v)MACsecSAI in SYNCD container. (v)MACsecSAI is a serial of APIs that are defined to communicate with the real SecY. In virtual switch, the SecY is Linux MACsec driver and vMACsecSAI will use the ip commands to manage them. But in real switch, the SecY is MACsec ASIC and these APIs will be provided by the vendor of MACsec ASIC.
```
[the remark this answers is not in the document]
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md]: Done, please review it.

File: doc/macsec/MACsec_hld.md [hunk 19]
```diff
@@ -0,0 +1,613 @@
...
+                                                       ; In 0-255 range with 0 being the highest priority
+CIPHER_SUITE                = "GCM-AES-128" / "GCM-AES-256" / "GCM-AES-XPN-128" / "GCM-AES-XPN-256"
+                                                       ; The cipher suite for MACsec.
+                                                       ; Default GCM-AES-128 if this filed not exist
+PRIMARY_CAK                 = 32HEXDIG / 64HEXDIG      ; Primary Connectivity Association Key
+PRIMARY_CKN                 = 64HEXDIG                 ; Primary CAK Name
+FALLBACK_CAK                = 32HEXDIG / 64HEXDIG      ; Fallback Connectivity Association Key
+FALLBACK_CKN                = 64HEXDIG                 ; Fallback CAK Name
+POLICY                      = 0/1/2/3                  ; MACsec policy. 0: Not security,
+                                                       ; 1: Only integrity checking, 3: encrypt traffic
```
[the remark this answers is not in the document]
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md]: Done, please review it.

File: doc/macsec/MACsec_hld.md [hunk 20]
```diff
@@ -0,0 +1,613 @@
...
+```
+
+### 3.2 App DB
+
+Fellowing new tables would be introduced to specify the MACsec parameters, SCs and SAs at the target port.
+
+#### 3.2.1 MACsec Port Table
+
+``` rfc5234
+"MACSEC_PORT"|{{port_name}}
```
[the remark this answers is not in the document]
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md]: Done, please review it.

File: doc/macsec/MACsec_hld.md (hunk 1 above)
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md]: Done, please review it.

File: doc/macsec/MACsec_hld.md (hunk 2 above)
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md]: Done, please review it.

File: doc/macsec/MACsec_hld.md (hunk 3 above)
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md]: Done, please review it.

File: doc/macsec/MACsec_hld.md (hunk 4 above)
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md]: Done, please review it.

File: doc/macsec/MACsec_hld.md (hunk 5 above)
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md]: Done, please review it.

File: doc/macsec/MACsec_hld.md (hunk 6 above)
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md]: Yes, we need to finally collect SA stats from counter DB when we delete it.

File: doc/macsec/MACsec_hld.md (hunk 7 above)
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md]: Done, please review it.

File: doc/macsec/MACsec_hld.md (hunk 8 above)
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md]: Done, please review it.

File: doc/macsec/MACsec_hld.md (hunk 9 above)
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md]: Done, please review it.

File: doc/macsec/MACsec_hld.md (hunk 11 above)
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md]: Done, please review it.

File: doc/macsec/MACsec_hld.md [hunk 21]
```diff
@@ -0,0 +1,743 @@
...
+  - **SONiC MACsec Plugin** is a plugin of wpa_supplicant that follows the predefined APIs of wpa_supplicant. It is responsible for bi-directional conversion of the MACsec parameters and SONiC database entries.
+
+- The green means these components are in SWSS container. This container uses the SAI APIs to control the MACsec security entities(SecY) according to databases entries and to synchronize the statistics from SecY to COUNTERS_DB.
+  - **MACsecOrch** is a module of orchagent, that uses SAI APIs to manage the SecY according to messages from databases and synchronized the statistics of SecY to COUNTERS_DB.
+
+- The blue one is MACsecSAI in SYNCD container. MACsecSAI is a set of APIs that are defined to communicate with the SecY. In the virtual switch, the SecY is Linux MACsec driver and MACsecSAI will use the ip commands to manage them. But in the real switch, the SecY is the MACsec cipher chip and the implementation of MACsecSAI will be provided by the vendor of the cipher chip.
```
Comment (qiluo-msft) [inline doc/macsec/MACsec_hld.md:144]: >The blue one is MACsecSAI in SYNCD container [](start = 2, length = 44)

The MACsecSAI box in the diagram is confusing. All other boxes are representing processes. And audience keep asking how many syncd processes in your design. You may draw syncd1/sync2 process as boxes, and MACsecSAI/SAI as internal boxes (like the plugin).

File: doc/macsec/MACsec_hld.md [hunk 22]
```diff
@@ -0,0 +1,743 @@
...
+
+#### Phase IV
+
+- MACsec can support confidentiality offset setting
+
+## 2 Architecture Design
+
+This chapter shows the MACsec interface stack of virtual switch and real switch.
+
+Virtual switch use the Linux MACsec driver as the MACsec Security Entity(SecY) to support the functionality of MACsec and the SecY is imposed on the physical port.
```
Comment (qiluo-msft) [inline doc/macsec/MACsec_hld.md:124]: >Virtual switch [](start = 0, length = 14)

I know you are talking about SONiC. But it is easy to mislead to the virtual switch between VMs  on a physical server. And in SONiC terminology, we already have a 'Virtual Switch SAI', I think your term means total different thing.

File: doc/macsec/MACsec_hld.md (hunk 21 above)
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md]: > > The blue one is MACsecSAI in SYNCD container [](start = 2, length = 44)
> 
> The MACsecSAI box in the diagram is confusing. All other boxes are representing processes. And audience keep asking how many syncd processes in your design. You may draw syncd1/sync2 process as boxes, and MACsecSAI/SAI as internal boxes (like the plugin).

Fixed, please review it.

File: doc/macsec/MACsec_hld.md (hunk 22 above)
Comment (Pterosaur) [reply doc/macsec/MACsec_hld.md]: > > Virtual switch [](start = 0, length = 14)
> 
> I know you are talking about SONiC. But it is easy to mislead to the virtual switch between VMs on a physical server. And in SONiC terminology, we already have a 'Virtual Switch SAI', I think your term means total different thing.

I discussed with Guohan and change it to SAI virtual switch, please review whether it makes sense.

**Comment by caizhenghui** [comment]
Juniper Networks has already implemented very similar Macsec support using wpa_supplicant, and want to contribute to the community on this feature. We believe from our work, we can accelerate development on the following items.

Enable wpa_supplicant supporting different Ciphersuite: GCM-AES-128, GCM-AES-256, GCM-AES-XPN-128 and GCM-AES-XPN-256
XPN feature support in wpa_supplicant.
CLI commands to configure MACsec (other than enable/disable)
Using libnl to pass macsec messages instead of ip command.

We can discuss in detail during HLD review.

**Comment by lguohan** [comment]
@caizhenghui, how to contact you let's have some discussion first?

**Comment by caizhenghui** [comment]
Hi, Guohan

Please use my company address zcai@juniper.net<mailto:zcai@juniper.net>..
We can have offline discussion.

Thanks

Zhenghui

File: doc/macsec/MACsec_hld.md [hunk 23]
```diff
@@ -0,0 +1,788 @@
...
+```
+
+#### 3.1.2 Port Table
+
+``` rfc5234
+PORT|{{port_name}}
+    "macsec":{{profile}} (OPTIONAL)
+
+; Defines schema for MACsec profile configuration attributes
+key                         = PORT:name               ; Interface name
```
Comment (heidinet2007) [inline doc/macsec/MACsec_hld.md:208]: For PortChannel support, should the physical interface name or the PortChannel name be specified here?

File: doc/macsec/MACsec_hld.md [hunk 24]
```diff
@@ -0,0 +1,788 @@
...
+
+This section describes the SONiC requirements for MACsec primary feature.
+
+At a high level the following should be supported:
+
+#### Phase I
+
+- MACsec can be enabled at a specified [port](https://github.com/Azure/SONiC/wiki/Configuration#port)
+- MACsec can co-work with the [port channel](https://github.com/Azure/SONiC/wiki/Configuration#port-channel)
+- Support Cipher: GCM-AES-128 and GCM-AES-256
```
Comment (heidinet2007) [inline doc/macsec/MACsec_hld.md:102]: How about  802.1Q subinterface?

Comment (lguohan) [reply doc/macsec/MACsec_hld.md:102]: no plan

File: doc/macsec/MACsec_hld.md (hunk 23 above)
Comment (lguohan) [reply doc/macsec/MACsec_hld.md:207]: physical interface name.

**Comment by qbdwlr** [comment]
As part of this MACsec effort, has there been any consideration for including a modified version of tcpdump that is capable of parsing the EAPOL exchange?

**Comment by Pterosaur** [comment]
> As part of this MACsec effort, has there been any consideration for including a modified version of tcpdump that is capable of parsing the EAPOL exchange?

Hi Ann, I believe Wireshark can parse the EAPOL traffic so right now, we don't have plan to modify tcpdump.

**Pull request merged**